File: internal/core/trail/domain/wal.go
package domain

import "time"

type WalTransactionResult struct {
	ID          uint64            `json:"id"`
	SourceSlot  string            `json:"source_slot"`
	SourceDb    string            `json:"source_db"`
	Xid         int64             `json:"xid"`
	CommitLSN   *string           `json:"commit_lsn,omitempty"`
	CommitTS    time.Time         `json:"commit_ts"`
	ChangeCount int32             `json:"change_count"`
	IngestedAt  time.Time         `json:"ingested_at"`
	Changes     []WalChangeResult `json:"changes"`
}

type WalChangeResult struct {
	ID             uint64    `json:"id"`
	TransactionID  uint64    `json:"transaction_id"`
	ChangeSeqInTxn int32     `json:"change_seq_in_txn"`
	SchemaName     string    `json:"schema_name"`
	TableName      string    `json:"table_name"`
	Op             string    `json:"op"`
	ChangedColumns []string  `json:"changed_columns"`
	ForwardDMLSQL  string    `json:"forward_dml_sql"`
	ReverseDMLSQL  string    `json:"reverse_dml_sql"`
	UndoStatus     string    `json:"undo_status"`
	CreatedAt      time.Time `json:"created_at"`
}

type PaginatedResponse[T any] struct {
	Data       []T `json:"data"`
	Page       int `json:"page"`
	PageSize   int `json:"page_size"`
	TotalCount int `json:"total_count"`
	TotalPages int `json:"total_pages"`
}

File: packages/frontend/src/api/wal.ts
import { apiFetch } from './client'

export interface Paginated<T> {
  data: T[]
  page: number
  page_size: number
  total_count: number
  total_pages: number
}

export interface WalTransaction {
  id: number
  source_slot: string
  source_db: string
  xid: number
  commit_lsn?: string
  commit_ts: string
  change_count: number
  ingested_at: string
  changes: WalChange[]
}

export interface WalChange {
  id: number
  transaction_id: number
  change_seq_in_txn: number
  schema_name: string
  table_name: string
  op: 'I' | 'U' | 'D'
  changed_columns: string[]
  forward_dml_sql: string
  reverse_dml_sql: string
  undo_status: string
  created_at: string
}

export function listWalTransactions(
  page = 1,
  pageSize = 20,
): Promise<Paginated<WalTransaction>> {
  return apiFetch<Paginated<WalTransaction>>(
    `/wal/transactions?page=${page}&page_size=${pageSize}`,
  )
}
